### PR TITLE
style context added to roc plot

### LINKED ISSUE #243: Coloring scheme
We can come up with a unique/prettier coloring scheme for our plots.
This can be unified across the different plots and we can also add a @Made by ploomber tag.
For instance, gigi plot on matplotlib (looks nicer than the default one)

### PR DESCRIPTION
**Pull request by yafimvo**
## Describe your changes
1. Style context added to control plots styles
2. `apply_theme` decorator added to plot functions
3. Base images updated
4. Modified existing plot hardcoded style

## Issue ticket number and link
Closes #243  

## How does it work?
Context style controls the visual style of our matplotlib plots in one place using the `rcParams`.
Currently, the default style is an "L" shaped chart (only left and bottom borders)  and a monochromatic coloring (tints of one color) based on the material UI color palette. We can modify both with the`cmap_style` and `ax_style` parameters.
* `cmap_style` : ["monochromatic", "gradient"]
* `ax_style` : ["no_frame", "frame"]

For example:
Let's style `_silhouette_analysis_one_model` which generates a silhouette plot for one value of n_cluster.

the default style is:

```python
@apply_theme()
def _silhouette_analysis_one_model(..., ax =None):
    # plot implementation
```
result:
image

the default style is:
```python
@apply_theme(cmap_style="gradient")
def _silhouette_analysis_one_model(..., ax =None):
    # plot implementation
```
result:
image

We can override theme by adding custom logic within the method, for example: adding a grid and right border.
```python
@apply_theme(cmap_style="gradient")
def _silhouette_analysis_one_model(..., ax =None):
    # plot implementation
    ax.spines['right'].set_visible(True)
    ax.grid(True)
```
image

## Checklist before requesting a review

- [X] I have performed a self-review of my code
- [X] I have added thorough [tests](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#testing) (when necessary).
- [X] I have added the right documentation in the [docstring](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#documenting-changes-and-new-features) and [changelog](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#changelog) (when needed)

----
:books: Documentation preview :books:: https://sklearn-evaluation--279.org.readthedocs.build/en/279/

### CODE REVIEW

**Comment by edublancas** [comment]
looks like this is still wip so I marked it as draft

**Comment by yafimvo** [comment]
@edublancas I added a decorator and context manager to style the plots and applied them to the ROC curve.

Let me know if the style is OK with you so I can update the tests and modify more plots.

[Here is an example of the ROC curve](https://sklearn-evaluation--279.org.readthedocs.build/en/279/api/plot.html#roc)

**Comment by edublancas** [comment]
looks good!

this can be easily applied to many line plots; but there are others that might need tweaks. for example, confusion matrix. how are you thinking of that one?

**Comment by yafimvo** [comment]
> this can be easily applied to many line plots; but there are others that might need tweaks. for example, confusion matrix. how are you thinking of that one?

I added some logic to the decorator and the context so we can control the applied theme, whether it's colors, ax settings, or any other specific configuration we want.

We can apply it on cm as follows:

```python
@apply_theme(ax_style=False)
def plot(self, ax=None, **theme):
    # custom plot creation logic
```

Since we're passing the `cmap` parameter to `_plot_cm` and calculating its value sometimes in the `__init__` method and sometimes in the `plot` method, and sometimes in the `_matrix.add`; I added this `**theme` parameter so we can easily extract the `cmap` from the "decorated" function without adding an extra logic and breaking current stuff.

@edublancas what do you think?

**Comment by edublancas** [comment]
> ; I added this **theme parameter so we can easily extract the cmap from the "decorated" function without adding an extra logic and breaking current stuff.

I'm not following how this `**theme` works.

**Comment by yafimvo** [comment]
> I'm not following how this `**theme` works.

It allows us to pass the `cmap` parameter via `theme` and use it in [`def _plot_cm`](https://github.com/ploomber/sklearn-evaluation/blob/6aadc171dc9f6ffa11c1c9fd62e96848417ed9a0/src/sklearn_evaluation/plot/classification.py#L288) without breaking the current inner API

```python
im = ax.imshow(cm, interpolation="nearest", cmap=cmap)
ax.figure.colorbar(im)
```

I can remove this `theme` parameter and apply cmap in the context manager, but I will have to change these functions:

* [def _plot_cm](https://github.com/ploomber/sklearn-evaluation/blob/6aadc171dc9f6ffa11c1c9fd62e96848417ed9a0/src/sklearn_evaluation/plot/classification.py#L284)

* [sklearn_evaluation.plot._matrix.add](https://github.com/ploomber/sklearn-evaluation/blob/6aadc171dc9f6ffa11c1c9fd62e96848417ed9a0/src/sklearn_evaluation/plot/_matrix.py#L36)

and some more places that use `default_heatmap`

**Comment by edublancas** [comment]
I think let's remove `**theme` for now. I think we can later offer an API to customize the plots but adding generic keyword arguments like `**theme` will be confusing for users.

File: src/sklearn_evaluation/plot/feature_ranking.py [hunk 1]
```diff
@@ -196,7 +198,11 @@ def feature_ranks_custom_algorithm(self, ranks):
         self._draw()
         return self.ax
 
+    def _set_default_colors(self, cmap):
```
Comment (edublancas) [inline src/sklearn_evaluation/plot/feature_ranking.py:201]: why is this needed? looks like it'll edit matplotlib's globals

File: src/sklearn_evaluation/plot/feature_ranking.py [hunk 2]
```diff
@@ -394,7 +404,8 @@ def _draw(self):
 
         # Draw the heatmap
         data = np.ma.masked_where(mask, self.ranks_)
-        mesh = self.ax.pcolormesh(data, cmap=self.colormap, vmin=-1, vmax=1)
+        # mesh = self.ax.pcolormesh(data, cmap=self.colormap, vmin=-1, vmax=1)
```
Comment (edublancas) [inline src/sklearn_evaluation/plot/feature_ranking.py:407]: can we remove this commented line?

File: src/sklearn_evaluation/plot/grid_search.py [hunk 3]
```diff
@@ -202,12 +201,17 @@ def _grid_search_single(grid_scores, change, subset, kind, ax, sort, params):
     ax.set_title("Grid search results")
     ax.set_ylabel("Mean score")
     ax.set_xlabel(change)
-    ax.legend(loc="best")
+    ax.legend()
```
Comment (edublancas) [inline src/sklearn_evaluation/plot/grid_search.py:204]: Im guessing this was removed since best is the default value?

File: src/sklearn_evaluation/plot/matplotlib/bar.py [hunk 4]
```diff
@@ -41,9 +44,9 @@ def horizontal(values, labels=None, sort=True, error=None, ax=None):
     y_pos = np.arange(len(values))
 
     if error is None:
-        ax.barh(y_pos, values)
+        ax.barh(y_pos, values, edgecolor="#fff")
     else:
-        ax.barh(y_pos, values, xerr=error)
+        ax.barh(y_pos, values, xerr=error, edgecolor="#fff")
```
Comment (edublancas) [inline src/sklearn_evaluation/plot/matplotlib/bar.py:49]: is there an alternative API that would set this from the `apply_theme` decorator? or do you think it's better to set it here

File: src/sklearn_evaluation/plot/pca.py [hunk 5]
```diff
@@ -114,7 +115,8 @@ def _plot_generic(n_components, principal_components, ax):
     for pc1 in range(n_components):
         for pc2 in range(pc1 + 1, n_components):
             ax[ax_ind].scatter(
-                principal_components[:, pc1], principal_components[:, pc2], s=50
+                principal_components[:, pc1], principal_components[:, pc2],
+                alpha=.6, edgecolor="face", s=50
```
Comment (edublancas) [inline src/sklearn_evaluation/plot/pca.py:119]: same as my previous comment. is it better to set the edgecolor property inline rather than at the apply_theme level?

File: src/sklearn_evaluation/plot/precision_recall.py [hunk 6]
```diff
@@ -19,7 +20,7 @@
 def _set_ax_settings(ax, name):
     ax.set_title(name)
     ax.set_xlim([0.0, 1.0])
-    ax.set_ylim([0.0, 1.05])
+    ax.set_ylim([0.0, 1.0])
```
Comment (edublancas) [inline src/sklearn_evaluation/plot/precision_recall.py:23]: any reason to change this?

**Review by edublancas — CHANGES_REQUESTED** [review]
looks awesome! added a few minor comments.

also, the grid in many plots is gone (e.g., elbow curve). let's bring it back since it helps a lot with visualization. whenever there is a numeric scale in the X or Y axis, having a grid helps users see where the bar/line/dot is

**Comment by edublancas** [comment]
also, please briefly document how to use this in the developer's guide: https://sklearn-evaluation.ploomber.io/en/latest/reference/developer.html

File: src/sklearn_evaluation/plot/precision_recall.py (hunk 6 above)
Comment (yafimvo) [reply src/sklearn_evaluation/plot/precision_recall.py:23]: This adds a small margin at the top. removed this since we don't have an upper border.

Let me know which one is better

image

image

File: src/sklearn_evaluation/plot/grid_search.py (hunk 3 above)
Comment (yafimvo) [reply src/sklearn_evaluation/plot/grid_search.py:204]: yes

File: src/sklearn_evaluation/plot/pca.py (hunk 5 above)
Comment (yafimvo) [reply src/sklearn_evaluation/plot/pca.py:119]: According to [matplotlib api](https://matplotlib.org/stable/tutorials/introductory/customizing.html#the-default-matplotlibrc-file) docs it's possible using these attributes:

```
#scatter.marker: o         # The default marker type for scatter plots.
#scatter.edgecolors: face  # The default edge colors for scatter plots.
```

I tried
```
plt.rcParams['scatter.edgecolors'] = 'face'
```

But for some reason, it doesn't work. btw, changing `scatter.marker` works.

File: src/sklearn_evaluation/plot/matplotlib/bar.py (hunk 4 above)
Comment (yafimvo) [reply src/sklearn_evaluation/plot/matplotlib/bar.py:49]: solved with this:
```
plt.rcParams['patch.edgecolor'] = '#fff'
plt.rcParams['patch.force_edgecolor'] = True
```

File: src/sklearn_evaluation/plot/feature_ranking.py (hunk 1 above)
Comment (yafimvo) [reply src/sklearn_evaluation/plot/feature_ranking.py:201]: Good catch! removed it.

File: src/sklearn_evaluation/plot/style.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+        return wrapper_func
+    return decorator
+
+
+def _set_default_plot_colors(cmap_style):
+    material_ui_colors = get_color_palette()
+    plt.rcParams["axes.prop_cycle"] = mpl.cycler(color=material_ui_colors)
+
+    if cmap_style == "gradient":
```
Comment (tonykploomber) [inline src/sklearn_evaluation/plot/style.py:122]: Should we print some warning message if the input `cmap_style` is not supported?

Like:
```
@apply_theme(cmap_style="odd_color")
def plot_bar_with_cmap(cmap=None):
...
```

It will print
```
Currently we only support `monochromatic` and `gradient` in cmap_style
```

File: src/sklearn_evaluation/plot/classification.py [hunk 8]
```diff
@@ -30,9 +31,11 @@ def plot(self, ax=None):
         if ax is None:
             _, ax = plt.subplots()
 
+        cmap = default_cmap()
+
         _plot_cm(
             self.cm,
-            cmap=default_heatmap(),
```
Comment (tonykploomber) [inline src/sklearn_evaluation/plot/classification.py:35]: nit: I think we can keep this 
```
        _plot_cm(
            self.cm,
            cmap=default_heatmap(),
```

File: src/sklearn_evaluation/plot/precision_recall.py (hunk 6 above)
Comment (edublancas) [reply src/sklearn_evaluation/plot/precision_recall.py:23]: the top one looks better

File: src/sklearn_evaluation/plot/pca.py (hunk 5 above)
Comment (edublancas) [reply src/sklearn_evaluation/plot/pca.py:120]: alright, if it doesn't work, let's harcode it then

File: docs/reference/developer.md [hunk 9]
```diff
@@ -242,6 +242,126 @@ def my_plotting_function(y_true, y_pred, ax=None):
     pass
 ```
 
+### Styling plots
+Using our context style feature we can change matplotlib `rcParams` and easily manipulate the overall visual style of our plots. The default style features a frame with borders only on the left and bottom, and a monochromatic color scheme of one color from the material UI color palette. However, we have the ability to adjust both the cmap and the style of the chart itself using the parameters 'cmap_style' and 'ax_style'.
+
```
Comment (neelasha23) [inline docs/reference/developer.md:247]: `only on the left and bottom **of the plot**`

File: docs/reference/developer.md [hunk 10]
```diff
@@ -242,6 +242,126 @@ def my_plotting_function(y_true, y_pred, ax=None):
...
+def plot_bar_with_cmap(cmap=None):
+    data = [[30, 25, 50, 20], [40, 23, 51, 17], [30, 25, 50, 20]]
+
+    X = np.arange(4)
+    fig = plt.figure()
+    ax = fig.add_axes([0, 0, 1, 1])
+
+    offset = 0
+    for i in range(len(data)):
+        color = mpl.colormaps.get_cmap(cmap)(float(i) / len(data))
```
Comment (neelasha23) [inline docs/reference/developer.md:355]: We can also add example that passes `ax_style` as parameter to the decorator, and one where we pass both `ax_style` and `cmap_style`

**Comment by edublancas** [comment]
i fixed the docs in master, please rebase

File: src/sklearn_evaluation/plot/classification.py (hunk 8 above)
Comment (idomic) [reply src/sklearn_evaluation/plot/classification.py:35]: @yafim please address

Comment (yafimvo) [reply src/sklearn_evaluation/plot/classification.py:35]: We updated the default cmap with new colors, so using this function will return the old coloring. I kept this function since we might want to go back to the old coloring or toggle between them.

**Comment by yafimvo** [comment]
@idomic fixed all rejects and rebased

File: src/sklearn_evaluation/plot/feature_ranking.py [hunk 11]
```diff
@@ -197,6 +199,7 @@ def feature_ranks_custom_algorithm(self, ranks):
         return self.ax
 
 
+@apply_theme()
```
Comment (edublancas) [inline src/sklearn_evaluation/plot/feature_ranking.py:202]: both this one and the 2D version are no longer displaying the docs correctly: https://sklearn-evaluation--279.org.readthedocs.build/en/279/api/plot.html#rank1d

perhaps it'll be fixed if we apply the decorator to RankD or if we apply it to the method that creates the plot?

**Review by edublancas — CHANGES_REQUESTED** [review]
just a minor comment

**Review by edublancas — APPROVED** [review]
